docs/Customization/Keys-Traits-Configs.rst

```diff
@@ -0,0 +1,95 @@
+    :language: scala
+    :start-after: DOC include start: GCD lazy trait
+    :end-before: DOC include end: GCD imp trait
+
+These traits are added to the default ``Top`` in Chipyard.
+
+.. literalinclude:: ../../generators/example/src/main/scala/Top.scala
+    :language: scala
+    :start-after: DOC include start: Top
+    :end-before: DOC include end: Top
+
+Mixins
+------
+
+Mixins set the keys to a non-default value. Together, the collection of Mixins which define a configuration generate the values for all the keys used by the generator.
+
+For example, the ``WithGCDMixin`` is parameterized by the type of GCD widget you want to instantiate. When this mixin is added to a config, the ``GCDKey`` is set to a instance of ``GCDParams``, informing the previously mentioned traits to instantiate and connect the GCD widget appropriately.
+
+.. literalinclude:: ../../generators/example/src/main/scala/ConfigMixins.scala
+    :language: scala
+    :start-after: DOC include start: GCD mixin
+    :end-before: DOC include end: GCD mixin
+
+We can use this mixin when composing our configs.
+
+.. literalinclude:: ../../generators/example/src/main/scala/RocketConfigs.scala
+    :language: scala
+    :start-after: DOC include start: GCDTLRocketConfig
+    :end-before: DOC include end: GCDTLRocketConfig
+
+
+BuildTop
+--------
+
+The ``BuildTop`` key is special, because sometimes, we need to instantiate TestHarness modules to interface with a custom widget. The ``BuildTop`` key provides a function which can call some method of the Top to instantiate these TestHarness modules. Since the ``BuildTop`` key is called from the TestHarness, these modules will appear in the TestHarness. The config system also lets the ``BuildTop`` key look recursively into previous definitions of itself. This enables composability of the Top configurations.
+
+For example, conside a config that contains the mixins ``WithGPIO ++ WithTSI``. We need to both instantiate the TSI serial adapter, and tie off the GPIO pins.
+
+.. literalinclude:: ../../generators/example/src/main/scala/ConfigMixins.scala
+    :language: scala
+    :start-after: DOC include start: tsi mixin
+    :end-before: DOC include end: tsi mixin
+
+
+.. literalinclude:: ../../generators/example/src/main/scala/ConfigMixins.scala
+    :language: scala
+    :start-after: DOC include start: gpio mixin
+    :end-before: DOC include end: gpio mixin
+
+The leftmost mixin is ``WithGPIO``, so at elaboration time, the ``BuildTop`` key is set according to that mixin. This mixin looks up the previously set value of ``BuildTop`` by ``up(BuildTop, site)`` and calls that function to get the Top  that would return. In this configuration, the previous ``BuildTop`` is set by ``WithTSI``. Thus after complete evaluation, the TSI ports are connected by the ``BuildTop`` function defined in ``WithTSI``, and the GPIO ports are connected by the ``BuildTop`` function defined in ``WithGPIO``.
```

Comment:
This section might be a bit confusing if people know the order of mixin evaluation. First the `WithTSI` would be evaluated then the `WithGPIO`. So maybe it would be better to explain it that order.

I would also change the wording of the "We need to both instantiate the TSI serial adapter, and tie off the GPIO pins" sentence to mention that the `WithGPIO` instantiates X GPIO ports and ties them off.